src/main/sensors/gyroanalyse.c

```diff
@@ -46,58 +46,85 @@
-#define FFT_RESOLUTION            ((float)FFT_SAMPLING_RATE_HZ / FFT_WINDOW_SIZE)
 // following bin must be at least 2 times previous to indicate start of peak
 #define FFT_MIN_BIN_RISE          2
 
 // lowpass frequency for smoothing notch centre point
 #define DYN_NOTCH_SMOOTH_FREQ_HZ  60
 // notch centre point will not go below this, must be greater than cutoff, mid of bottom bin
 #define DYN_NOTCH_MIN_CENTRE_HZ   125
-// maximum notch centre frequency limited by Nyquist
-#define DYN_NOTCH_MAX_CENTRE_HZ   (FFT_SAMPLING_RATE_HZ / 2)
 // lowest allowed notch cutoff frequency
 #define DYN_NOTCH_MIN_CUTOFF_HZ   105
 // we need 4 steps for each axis
 #define DYN_NOTCH_CALC_TICKS      (XYZ_AXIS_COUNT * 4)
 
+static bool FAST_RAM_ZERO_INIT gyroAnalyseInitialized;
+static uint16_t FAST_RAM_ZERO_INIT fftSamplingRateHz;
+// centre frequency of bandpass that constrains input to FFT
+static uint16_t FAST_RAM_ZERO_INIT fftBpfHz;
+// Hz per bin
+static float FAST_RAM_ZERO_INIT    fftResolution;
+// maximum notch centre frequency limited by Nyquist
+static uint16_t FAST_RAM_ZERO_INIT dynNotchMaxCentreHz;
+static uint8_t  FAST_RAM_ZERO_INIT fftBinOffset;
+
 // Hanning window, see https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window
 static FAST_RAM_ZERO_INIT float hanningWindow[FFT_WINDOW_SIZE];
 static FAST_RAM_ZERO_INIT float dynamicNotchCutoff;
 
-void gyroDataAnalyseInit(void)
+// Making this NOINLINE saves a few bytes to help F3's compile
+// It's only called once at initialization so no performance concerns
+NOINLINE void gyroDataAnalyseInit(uint32_t targetLooptimeUs)
 {
+    const int gyroLoopRateHz = lrintf((1 / (float)targetLooptimeUs) * 1e6);
+
+    // If we get at least 3 samples then use the default FFT sample frequency
+    // otherwise we need to calculate a FFT sample frequency to ensure we get 3 samples (gyro loops < 4K)
+    fftSamplingRateHz = MIN((gyroLoopRateHz / 3), FFT_SAMPLING_RATE_HZ);
+
+    fftBpfHz = fftSamplingRateHz / 4;
+    fftResolution = (float)fftSamplingRateHz / FFT_WINDOW_SIZE;
+    dynNotchMaxCentreHz = fftSamplingRateHz / 2;
+
+    // Calculate the FFT bin offset to try and get the lowest bin used
+    // in the center calc close to 90hz
+    // > 1333hz = 1, 889hz (2.67K) = 2, 666hz (2K) = 3
+    fftBinOffset = MAX(1, lrintf(90 / fftResolution - 1.5f));
+
     for (int i = 0; i < FFT_WINDOW_SIZE; i++) {
         hanningWindow[i] = (0.5f - 0.5f * cos_approx(2 * M_PIf * i / (FFT_WINDOW_SIZE - 1)));
     }
 
     dynamicNotchCutoff = (100.0f - gyroConfig()->dyn_notch_width_percent) / 100;
+    gyroAnalyseInitialized = true;
 }
 
 void gyroDataAnalyseStateInit(gyroAnalyseState_t *state, uint32_t targetLooptimeUs)
 {
     // initialise even if FEATURE_DYNAMIC_FILTER not set, since it may be set later
+
+    // Perform the main initialization on the first gyro sensor init. We need
+    // the target loop time and that's only available after the sensor is initialized.
+    if (!gyroAnalyseInitialized) {
```

Comment:
I don't like this as well as elaborate comment. Can't you find another place where this could be called? Let it be called twice, but get rid of `init` flag

Reply:
Well it's painful in general.  The problem is that the initialization was originally run from `gyroInit()` but we now need the `targetLooptime` which isn't available until after the sensors are initialized in `gyroInitSensor()` (which may be called twice). This init then called `gyroDataAnalyseStateInit()` which relies on values calculated by `gyroDataAnalyseInit()` (with the `rargetLooptime` passed to it).  So I don't see any other good place to call it since `targetLooptime` isn't set until the first sensor initialization.

There's no technical problem with running the `gyroDataAnalyseInit()` twice other than it's unnecessary - hence the reason for the initialized flag and the explanation.

I suspect that if I changed it to just run twice I'll need a comment explaining that as well otherwise somebody will come along later and wonder why there's a bug with the init running twice for dual-gyro.